examples/hello_debugger.cpp

```diff
@@ -84,24 +119,45 @@ class Debugger {
   std::mutex mutex;
   int line = 1;
   std::unordered_set<int> breakpoints;
+  std::atomic<bool> terminate{false};
+  std::atomic<bool> running{true};
+  ::Event canRun;
 };
 
 Debugger::Debugger(const EventHandler& onEvent) : onEvent(onEvent) {}
 
 void Debugger::run() {
-  std::unique_lock<std::mutex> lock(mutex);
-  for (int i = 0; i < numSourceLines; i++) {
-    auto l = ((line + i) % numSourceLines) + 1;
-    if (breakpoints.count(l)) {
-      line = l;
+  while (!terminate) {
```

Comment:
So, although you're right in that adding breakpoints while the debugger was 'running' would not break, I was very weary of hard-spinning. I feel like keeping this spin-free would still be nice.
Perhaps if we had an internal debugger state (`Running`, `Paused`), we could fix the break behaviour without burning a hole in the CPU? I think the public API would keep pretty much the same. Note, this would also simplify the atomic logic below (just use the mutex for synchronization).

Reply:
The reason I was looking at this is because I was thinking about my emulator, and how I'd implement the DAP debugger for it, and in my case, I would be running it's simulation frame over and over, like we do it games. So I figured it would be useful if this example debugger did a similar thing, especially since the step behaviour is to loop indefinitely. I'm not too worried about CPU usage since this is only an example, but if it is a concern, we could perhaps add a sleep(1ms) when m_running is true in the loop.

> Note, this would also simplify the atomic logic below (just use the mutex for synchronization

I'm all for simpler code, but only if it represents something similar to what one would do in a real application. If we do as you suggest, I think what you're saying is that we fake that we're actually running, so that when a breakpoint is set, and we're paused, we make it run. This would work, but is not representative of how a real application would do it, right?

Reply:
Okay, that seems like a fair argument.